age-plugin.md

```diff
@@ -0,0 +1,541 @@
+- An `error` command with three variants:
+  - `(error, identity IDENTITY_INDEX; MESSAGE)` - an error occured while using
+    this identity.
+    - Response is `(ok)`.
+  - `(error, stanza FILE_INDEX STANZA_INDEX; MESSAGE)` - an error occured while
+    using a specific stanza. This MUST NOT be used for unknown stanzas, only for
+    stanzas with an expected tag but that are otherwise invalid (indicating an
+    invalid age header).
+    - Response is `(ok)`.
+  - `(error, internal; MESSAGE)` - a general error occurred during unwrapping.
+    - Response is `(ok)`.
+
+Having assembled the full list of identities to use, and supported stanzas to
+trial-unwrap, the plugin enforces structural validity on both:
+- If there are unknown or invalid identity types, the plugin MUST return errors
+  and MUST NOT attempt to unwrap any file keys with otherwise-valid identities.
+- If any known stanza is structurally invalid, the plugin MUST return an error
+  for that stanza, and MUST NOT unwrap any stanzas with the same `FILE_INDEX`.
+  The plugin MAY continue to unwrap stanzas from other files.
+
+The plugin then chooses internally an order of identities to try, and sends
+commands to request additional help from the client / user, or store unwrapped
+file keys.
+
+The plugin is the controller of this phase; clients should not close the
+connection if, for example, a user fails to respond in the way the plugin wants
+for a particular request. Instead, the client returns `(fail)` to indicate this;
+the plugin is expected to try another recipient stanza or identity.
+
+When the plugin is able to determine whether a given file key can be unwrapped
+or not, it takes one of three actions:
+- If a stanza cannot be unwrapped that detectably should be unwrappable (e.g.
+  the stanza specifically identifies the recipient), the plugin sends
+  `error stanza`.
+  - TODO: Should this be a hard error (preventing that file from being unwrapped)?
+    Probably yes (we already assume we don't get 32-bit collisions for SSH tags).
+- If any recipient stanza with a given `FILE_INDEX` can be unwrapped, it sends
+  `file-key`.
+- If all known and valid stanzas for a given file cannot be unwrapped, and none
+  are expected to be unwrappable, the plugin does not send anything. That is,
+  file keys that cannot be unwrapped are implicit.
+
+These may be interleaved with other requests, and the client must cache the
+unwrapped file keys until this phase is complete. The plugin sends no more than
+one file key per file index in the original set of stanzas.
+
+- TODO: Should IPC errors (e.g. invalid command structure or argument types) be
+  separated from state machine errors (e.g. invalid identities)?
+  - Probably not; it should be possible to run the IPC protocol over e.g. TCP.
+- TODO: What about identities that are not unknown or invalid, but for which we
+  get an error while trying to use them? (e.g. YubiKey missing, invalid PIN,
+  cloud KMS rejection)
```

Comment:
Some of these errors, such as the user deciding to skip the YubiKey through a `confirm` dialog, or a KMS plugin that encounters a transient error but wants to give a chance to a fallback identity, should be soft errors that let the client continue on to the next identity. The plugin can implement these by optionally sending a `msg` and then returning without sending anything, like when no stanzas are expected to be unwrappable, as described above.

Reply:
Opened #32 for subsequent work.